### PR TITLE
Prompt as Controller

### PR DESCRIPTION
**Pull request by shargon**
- Clean code
- Now you can inject types in prompt methods

### CODE REVIEW

**Review by osmirnov — COMMENTED** [review]
I think to inject the container the very last resort because after that nobody can guess which deps the class actually use.

I see a couple times the question about simplification in prompt design. Right now it is a huge class that using partial C# feature to split the code among many files. Needless to say, it hurts it more than helps and injecting the container is one more patch to address design issues.
 
Most likely, Prompt class is gonna grow. I'd like to suggest the design we pioneered in Network: Commands + CommandHandlers. Having such command classes will help to see deps and Prompt class will serve just as delegator input and output.

@aboimpinto what do you think?

**Comment by shargon** [comment]
I can made controllers, and inject this modules inside Prompt

Take a look here @osmirnov 

https://github.com/shargon/neo-sharp/pull/38/files

File: src/NeoSharp.Application/Attributes/PromptCommandAttribute.cs [hunk 1]
```diff
@@ -136,48 +136,66 @@ internal void SetMethod(MethodInfo method)
         /// Convert string arguments to Method arguments
         /// </summary>
         /// <param name="args">Arguments</param>
+        /// <param name="injector">Injector</param>
         /// <returns>Return parsed arguments</returns>
-        public object[] ConvertToArguments(CommandToken[] args)
+        public object[] ConvertToArguments(CommandToken[] args, IContainer injector)
```
Comment (aboimpinto) [inline src/NeoSharp.Application/Attributes/PromptCommandAttribute.cs]: I think we can find a better way to do this. IContainer should never been injected.

File: src/NeoSharp.Application/Attributes/PromptCommandAttribute.cs [hunk 2]
```diff
@@ -136,48 +136,66 @@ internal void SetMethod(MethodInfo method)
...
+                        ret[x] = obj;
+                        injected.Add(x);
+                    }
+                }
+            }
 
             // Fill argument values
 
-            for (int x = 0, amax = Math.Min(args.Length, maxPars); x < amax; x++)
+            for (int xP = 0, xA = 0; xP < maxPars && xA < args.Length; xP++, xA++)
```
Comment (aboimpinto) [inline src/NeoSharp.Application/Attributes/PromptCommandAttribute.cs]: can we have better names? What xP and xA mean?

File: src/NeoSharp.Application/Client/Prompt.cs [hunk 3]
```diff
@@ -106,78 +89,37 @@ public partial class Prompt : IPrompt
...
-        /// <param name="rpcInit">Rpc server</param>
-        /// <param name="serializer">Binary serializer</param>
         /// <param name="blockchain">Blockchain</param>
-        /// <param name="blockPool">Block pool</param>
         /// <param name="walletManager"></param>
         /// <param name="vmFactory">VM Factory</param>
-        public Prompt(
+        public Prompt
+            (
+            IContainer container,
```
Comment (aboimpinto) [inline src/NeoSharp.Application/Client/Prompt.cs:86]: this is not solution ... IContainer should never been injected

**Review by aboimpinto — CHANGES_REQUESTED** [review]
This is over-engineering. Still believe that you are an amazing developer but this is over the top solution for an easy problem.

Instead of inject the IContainer, I think the solution is to use a known pattern like "Strategy" and inject a list of ICommands (some interface like this) where only one of the commands will be the one that process the written command.

Let's discuss this on the Issues page and  find a better solution.

**Comment by shargon** [comment]
Check now the controller system.

All is refactored except this (by now)

`Prompt[Usability]`

File: src/NeoSharp.Application/Client/Prompt.cs (hunk 3 above)
Comment (shargon) [reply src/NeoSharp.Application/Client/Prompt.cs:86]: I only be able to inject collection with generic types, what is the right way to register a collection of all types?

```csharp

 var promptHandlerTypes = typeof(IPromptController).Assembly
               .GetExportedTypes()
               .Where(t => t.IsClass && !t.IsAbstract && typeof(IPromptController).IsAssignableFrom(t))
               .ToArray();

            containerBuilder.RegisterSingleton(() => new PromptControllerFactory(promptHandlerTypes));

```

**Review by osmirnov — APPROVED** [review]
I think we are back on the right track!

_P.S.: How many branches do you work at the same time?_ 🤣

**Comment by shargon** [comment]
I will merge it, we can remove IContainer from prompt later, is only used now for create the controllers

**Pull request merged**